avoid unnecessary updates to CRDs and webhooks

steved opened the pull request:
## Problem Statement

Even with no changes to the service or secrets, we're seeing update calls to the CRD and webhook definitions.

## Related Issue

_N/A_

## Proposed Changes

Similar to controller-runtime's [CreateOrUpdate](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.20.3/pkg/controller/controllerutil/controllerutil.go#L327), check object equality before calling Update.

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [ ] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [ ] I ensured my PR is ready for review with `make reviewable`

Skarlso commented on pkg/controllers/crds/crds_controller.go:
```diff
@@ -227,7 +230,12 @@ func (r *Reconciler) updateCRD(ctx context.Context, req ctrl.Request) error {
 	}
-	return r.Update(ctx, &updatedResource)
+	if !equality.Semantic.DeepEqual(before, &updatedResource) {
+		defer log.Info("updated crd")
```
why defer log here? 🤔  Just log it?

steved replied:
err yeah got too cute and wanted it to log after the Update call, but we don't really want it if the error happens.

Skarlso commented on pkg/controllers/webhookconfig/webhookconfig.go:
```diff
@@ -189,7 +192,12 @@ func (r *Reconciler) updateConfig(ctx context.Context, cfg *admissionregistratio
+		defer log.Info("updated webhook config")
+		return r.Update(ctx, cfg)
+	}
+	defer log.Info("webhook config unchanged")
```
Again... why defer? :D

gusfcarvalho commented:
Sorry to jump here - can you please add the following test cases?

* When a CRD has no change, we see the resourceVersion not increasing (basically testing your feature)
* When a new certificate is generated, it is propagated and updated to the CRD (so we keep non regression)

steved commented:
> Sorry to jump here - can you please add the following test cases?
> 
> * When a CRD has no change, we see the resourceVersion not increasing (basically testing your feature)
> * When a new certificate is generated, it is propagated and updated to the CRD (so we keep non regression)
> 
> (same for the webhook config)

No problem! I've added a few more tests. I think with a real api server, my understanding is that we won't see resource version updates when the resource doesn't change?

gusfcarvalho commented:
Yeah! This is the behavior I wanted to capture!!

Skarlso commented:
/ok-to-test sha=3e6b0232d7d5713d7fa05d543c9dbaf114b20890
